diff --git a/services/surfaceflinger/SurfaceFlinger.cpp b/services/surfaceflinger/SurfaceFlinger.cpp
index db26808..9dde654 100644
--- a/services/surfaceflinger/SurfaceFlinger.cpp
+++ b/services/surfaceflinger/SurfaceFlinger.cpp
@@ -339,6 +339,25 @@ SurfaceFlinger::SurfaceFlinger(Factory& factory) : SurfaceFlinger(factory, SkipI
             SurfaceFlinger::primaryDisplayOrientation = DisplayState::eOrientationDefault;
             break;
     }
+
+    if(primaryDisplayOrientation == V1_1::DisplayOrientation::ORIENTATION_0) {
+        int sfRotation = property_get_int32("ro.sf.hwrotation", -1);
+        switch(sfRotation) {
+            case 0:
+                SurfaceFlinger::primaryDisplayOrientation = DisplayState::eOrientationDefault;
+                break;
+            case 90:
+                SurfaceFlinger::primaryDisplayOrientation = DisplayState::eOrientation90;
+                break;
+            case 180:
+                SurfaceFlinger::primaryDisplayOrientation = DisplayState::eOrientation180;
+                break;
+            case 270:
+                SurfaceFlinger::primaryDisplayOrientation = DisplayState::eOrientation270;
+                break;
+        }
+    }
+
     ALOGV("Primary Display Orientation is set to %2d.", SurfaceFlinger::primaryDisplayOrientation);
 
     mInternalDisplayPrimaries = sysprop::getDisplayNativePrimaries();
